# Automatidata project 
**Course 4 - The Power of Statistics**

You are a data professional in a data consulting firm, called Automatidata. The current project for their newest client, the New York City Taxi & Limousine Commission (New York City TLC) is reaching its midpoint, having completed a project proposal, Python coding work, and exploratory data analysis.

You receive a new email from Uli King, Automatidata’s project manager. Uli tells your team about a new request from the New York City TLC:**to analyze the relationship between fare amount and payment type.** A follow-up email from Luana includes your specific assignment: **to conduct an A/B test.** 

A notebook was structured and prepared to help you in this project. Please complete the following questions.


# Course 4 End-of-course project: Statistical analysis

In this activity, you will practice using statistics to analyze and interpret data. The activity covers fundamental concepts such as descriptive statistics and hypothesis testing. You will explore the data provided and conduct A/B and hypothesis testing.  
<br/>   

**The purpose** of this project is to demostrate knowledge of how to prepare, create, and analyze A/B tests. Your A/B test results should aim to find ways to generate more revenue for taxi cab drivers.

**Note:** For the purpose of this exercise, assume that the sample data comes from an experiment in which customers are randomly selected and divided into two groups: 1) customers who are required to pay with credit card, 2) customers who are required to pay with cash. Without this assumption, we cannot draw causal conclusions about how payment method affects fare amount.

**The goal** is to apply descriptive statistics and hypothesis testing in Python. The goal for this A/B test is to sample data and analyze whether there is a relationship between payment type and fare amount. For example: discover if customers who use credit cards pay higher fare amounts than customers who use cash.
  
*This activity has four parts:*

**Part 1:** Imports and data loading
* What data packages will be necessary for hypothesis testing?
               
               * numpy, pandas, from scipy import stats 

**Part 2:** Conduct EDA and hypothesis testing
* How did computing descriptive statistics help you analyze your data? 


  * In general, descriptive statistics are useful because they let you quickly explore and understand large amounts of data. 
   
               * The mean (average) gives us a general idea of the typical fare in the dataset. It’s useful for understanding the overall trend of the data. For example, when looking at taxi fares, the mean fare shows us the average amount customers are paying, providing a quick sense of the typical fare. However, the mean can be affected by very high or very low values (outliers), which is why it's often helpful to also consider the median and standard deviation for a fuller picture of the data.

            * Median: The median tells us the "middle" fare when the data is sorted. It’s useful because it isn’t affected by extreme values, offering a more accurate reflection of the typical fare when there are outliers.

            * Standard Deviation: The standard deviation shows how much the fares deviate (or vary) from the mean.If the standard deviation is small, it means the fares are mostly close to the mean (indicating consistency).If the standard deviation is large, it means the fares are more spread out, with some being much higher or lower than the mean (indicating more variation).In short, the standard deviation helps us understand how consistent or varied the fares are compared to the average fare.
          
          * Identifying Trends:When calculating the mean or median fare for both credit card and cash payments, we can compare these values to see if one payment method generally results in higher or lower fares. For example:If the mean fare for credit card payments is higher than for cash payments, this suggests that credit card users tend to pay more on average.Similarly, comparing the median fare for each payment type helps identify trends, especially when there are outliers, as the median is less affected by extreme values.
          
          * Checking for Outliers:To identify outliers, we compare the mean with the maximum and minimum values. Here's how this works:Maximum/Minimum Values: If the maximum or minimum values are far away from the mean, it suggests that there might be outliers. For example, if the average fare is 30, but the maximum fare is 1000 or the minimum fare is 1, these extreme values may not be typical and could be outliers.
      
         * Mean vs. Range:If there is a big difference between the mean and the maximum or minimum values, it suggests that most of the data points are close to the mean, but there are a few extreme values (either very high or very low) that are far from the mean. These extreme values could be outliers—values that don't follow the usual pattern of the data and might need further investigation.

* How did you formulate your null hypothesis and alternative hypothesis? 

      * H₀: There is no difference in the average fare amount between credit card and cash payments.
      * H₁: Customers who use a credit card pay higher fare amounts than those who pay with cash.


**Part 3:** Communicate insights with stakeholders

* What key business insight(s) emerged from your A/B test?

* What business recommendations do you propose based on your results?

<br/> 
Follow the instructions and answer the questions below to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work. 

# **Conduct an A/B test**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## PACE: Plan 

In this stage, consider the following questions where applicable to complete your code response:
1. What is your research question for this data project? Later on, you will need to formulate the null and alternative hypotheses as the first step of your hypothesis test. Consider your research question now, at the start of this task.


**Research Question: Does the payment type (credit card or cash) affect the taxi fare amount?**


*Complete the following steps to perform statistical analysis of your data:* 

### Task 1. Imports and data loading

Import packages and libraries needed to compute descriptive statistics and conduct a hypothesis test.

<details>
  <summary><h4><strong>Hint: </strong></h4></summary>

Before you begin, recall the following Python packages and functions that may be useful:

*Main functions*: stats.ttest_ind(a, b, equal_var)

*Other functions*: mean() 

*Packages*: pandas, stats.scipy

</details>

In [4]:
import pandas as pd
import numpy as np
from scipy import stats

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [5]:
# Load dataset into dataframe
taxi_data = pd.read_csv("2017_Yellow_Taxi_Trip_Data.csv", index_col = 0)

<img src="images/Analyze.png" width="100" height="100" align=left>

<img src="images/Construct.png" width="100" height="100" align=left>

## PACE: **Analyze and Construct**

In this stage, consider the following questions where applicable to complete your code response:
1. Data professionals use descriptive statistics for Exploratory Data Analysis. How can computing descriptive statistics help you learn more about your data in this stage of your analysis?


 **In general, descriptive statistics are useful because they let you quickly explore and understand large amounts of data.**
   
   **The mean (average)** gives us a general idea of the typical fare in the dataset. It’s useful for understanding the overall trend of the data. For example, when looking at taxi fares, the mean fare shows us the average amount customers are paying, providing a quick sense of the typical fare. However, the mean can be affected by very high or very low values (outliers), which is why it's often helpful to also consider the median and standard deviation for a fuller picture of the data.

 **Median:** The median tells us the "middle" fare when the data is sorted. It’s useful because it isn’t affected by extreme values, offering a more accurate reflection of the typical fare when there are outliers.

 **Standard Deviation:** The standard deviation shows how much the fares deviate (or vary) from the mean.If the standard deviation is small, it means the fares are mostly close to the mean (indicating consistency).If the standard deviation is large, it means the fares are more spread out, with some being much higher or lower than the mean (indicating more variation).In short, the standard deviation helps us understand how consistent or varied the fares are compared to the average fare.
          
 **Identifying Trends:** When calculating the mean or median fare for both credit card and cash payments, we can compare these values to see if one payment method generally results in higher or lower fares. For example:If the mean fare for credit card payments is higher than for cash payments, this suggests that credit card users tend to pay more on average.Similarly, comparing the median fare for each payment type helps identify trends, especially when there are outliers, as the median is less affected by extreme values.
          
**Checking for Outliers:** To identify outliers, we compare the mean with the maximum and minimum values. Here's how this works:Maximum/Minimum Values: If the maximum or minimum values are far away from the mean, it suggests that there might be outliers. For example, if the average fare is 30, but the maximum fare is 1000 or the minimum fare is 1, these extreme values may not be typical and could be outliers.
      
**Mean vs. Range:** If there is a big difference between the mean and the maximum or minimum values, it suggests that most of the data points are close to the mean, but there are a few extreme values (either very high or very low) that are far from the mean. These extreme values could be outliers—values that don't follow the usual pattern of the data and might need further investigation.


### Task 2. Data exploration

Use descriptive statistics to conduct Exploratory Data Analysis (EDA). 

<details>
  <summary><h4><strong>Hint: </strong></h4></summary>

Refer back to *Self Review Descriptive Statistics* for this step-by-step proccess.

</details>

**Note:** In the dataset, `payment_type` is encoded in integers:
*   1: Credit card
*   2: Cash
*   3: No charge
*   4: Dispute
*   5: Unknown



In [6]:
taxi_data.head(5)

VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
24870114          2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
35634249          1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
38942136          2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
30841670          2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   

           passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
24870114                 6           3.34           1                  N   
35634249                 1           1.80           1                  N   
106203690                1           1.00           1                  N   
38942136                 1           3.70           1                  N   
30841670                 1           4.37           1                  N   

           PULocationID  DOLocationID  payment_type  fare_amount  extra  \
24870114            100           231             1         13.0    0.0   
35634249            186            43             1         16.0    0.0   
106203690           262           236             1          6.5    0.0   
38942136            188            97             1         20.5    0.0   
30841670              4           112             2         16.5    0.5   

           mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
24870114       0.5        2.76           0.0                    0.3   
35634249       0.5        4.00           0.0                    0.3   
106203690      0.5        1.45           0.0                    0.3   
38942136       0.5        6.39           0.0                    0.3   
30841670       0.5        0.00           0.0                    0.3   

           total_amount  
24870114          16.56  
35634249          20.80  
106203690          8.75  
38942136          27.69  
30841670          17.80

In [9]:
taxi_data.shape

(22699, 17)

22699 rows and 17 columns

In [10]:
taxi_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22699 entries, 24870114 to 17208911
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   VendorID               22699 non-null  int64  
 1   tpep_pickup_datetime   22699 non-null  object 
 2   tpep_dropoff_datetime  22699 non-null  object 
 3   passenger_count        22699 non-null  int64  
 4   trip_distance          22699 non-null  float64
 5   RatecodeID             22699 non-null  int64  
 6   store_and_fwd_flag     22699 non-null  object 
 7   PULocationID           22699 non-null  int64  
 8   DOLocationID           22699 non-null  int64  
 9   payment_type           22699 non-null  int64  
 10  fare_amount            22699 non-null  float64
 11  extra                  22699 non-null  float64
 12  mta_tax                22699 non-null  float64
 13  tip_amount             22699 non-null  float64
 14  tolls_amount           22699 non-null  float

In [12]:
# Missing values
taxi_data.isnull().sum()


VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
RatecodeID               0
store_and_fwd_flag       0
PULocationID             0
DOLocationID             0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
dtype: int64

* No null values in any column

In [13]:
# summary statistics

taxi_data.describe()

VendorID  passenger_count  trip_distance    RatecodeID  \
count  22699.000000     22699.000000   22699.000000  22699.000000   
mean       1.556236         1.642319       2.913313      1.043394   
std        0.496838         1.285231       3.653171      0.708391   
min        1.000000         0.000000       0.000000      1.000000   
25%        1.000000         1.000000       0.990000      1.000000   
50%        2.000000         1.000000       1.610000      1.000000   
75%        2.000000         2.000000       3.060000      1.000000   
max        2.000000         6.000000      33.960000     99.000000   

       PULocationID  DOLocationID  payment_type   fare_amount         extra  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean     162.412353    161.527997      1.336887     13.026629      0.333275   
std       66.633373     70.139691      0.496211     13.243791      0.463097   
min        1.000000      1.000000      1.000000   -120.000000     -1.000000   
25%      114.000000    112.000000      1.000000      6.500000      0.000000   
50%      162.000000    162.000000      1.000000      9.500000      0.000000   
75%      233.000000    233.000000      2.000000     14.500000      0.500000   
max      265.000000    265.000000      4.000000    999.990000      4.500000   

            mta_tax    tip_amount  tolls_amount  improvement_surcharge  \
count  22699.000000  22699.000000  22699.000000           22699.000000   
mean       0.497445      1.835781      0.312542               0.299551   
std        0.039465      2.800626      1.399212               0.015673   
min       -0.500000      0.000000      0.000000              -0.300000   
25%        0.500000      0.000000      0.000000               0.300000   
50%        0.500000      1.350000      0.000000               0.300000   
75%        0.500000      2.450000      0.000000               0.300000   
max        0.500000    200.000000     19.100000               0.300000   

       total_amount  
count  22699.000000  
mean      16.310502  
std       16.097295  
min     -120.300000  
25%        8.750000  
50%       11.800000  
75%       17.800000  
max     1200.290000

**about fare_amount **

**Mean (13.02):**
* The average fare is approximately $13.02. However, the mean can be influenced by extreme values (outliers), such as very high or very low fares.

**Median (9.5):**
* The middle value of the fares is $9.50. Since the median is lower than the mean, it suggests that there may be some high outliers pulling the mean upward.

**Standard Deviation (13.243):**
* The mean is 13.02, and the standard deviation is 13.243.
 When the standard deviation is close to or larger than the mean, it suggests high variability in the data.
This means that fares are spread out widely, with many values far from the average fare of $13.02.

** Minimum Value (-$120):**
* Unusual Compared to Mean:
The mean fare is $13.02, which is reasonable for a typical taxi ride.
A negative fare is not only far from the mean but also logically incorrect since fares can't be negative.
This suggests a data error, such as a mistake during data entry, a refund recorded incorrectly, or an issue in data collection.
**Maximum Value ($999.99):**
* Extreme Compared to Mean:
The mean fare is $13.02, while the maximum is $999.99—this is almost 77 times the mean.
**25% (6.5):**
* 25% of fares are below $6.50, indicating that many fares are relatively low.

**75% (14.5):**
* 75% of fares are below $14.50. This shows that the majority of fares are clustered within a reasonable range, with only a few going beyond this value.

**Negative Minimum (-120):**
* A negative fare is not valid and should be flagged as a potential data error.

**High Maximum (999.99):**
* This could be a legitimate fare for an exceptionally long trip, but it should be checked for validity.

**Overall Insights**
* Most fares fall between $6.50 (25%) and $14.50 (75%), with a median of $9.50, showing that the majority of fares are affordable and consistent.
* The mean being higher than the median suggests the presence of outliers that skew the average upward.
The data contains anomalies, such as the negative fare, which may need to be cleaned or excluded during analysis.

**Next Steps**
**Data Cleaning:**
Remove or investigate the negative fares and unusually high fares.
**Further Analysis:**
Examine the distribution of the data (e.g., histogram) to confirm the skewness caused by outliers.
Consider using robust measures like the median and interquartile range (IQR) to summarize the data if the outliers significantly affect the analysis.

**about payment_type**

* In the dataset, payment_type is encoded in integers:

1: Credit card
2: Cash
3: No charge
4: Dispute
5: Unknown

* For categorical variables( payment_type ), summary statistics such as mean, standard deviation (std), minimum, and maximum do not make sense because these calculations are meant for numerical data, not categories.
* Let us Use Python's value_counts() function to determine how often each payment type appears.


In [14]:
taxi_data['payment_type'].value_counts()

1    15265
2     7267
3      121
4       46
Name: payment_type, dtype: int64

**The counts show the frequency of each payment type:**

* Credit Card (1): 15,265 trips, the most common payment method.
* Cash (2): 7,267 trips, the second most frequent method.
* No Charge (3): 121 trips, likely complimentary or promo rides.
* Dispute (4): 46 trips, where payments were contested.( a "dispute" often means the payment is in question, and it might not be finalized until the issue is resolved.)

Insights:
    * Most trips are paid with credit cards, followed by cash.
    * Rare scenarios like "No Charge" and "Disputes" might need further investigation.

In [16]:
taxi_data.describe(include='all')

VendorID   tpep_pickup_datetime  tpep_dropoff_datetime  \
count   22699.000000                  22699                  22699   
unique           NaN                  22687                  22688   
top              NaN  07/03/2017 3:45:19 PM  10/18/2017 8:07:45 PM   
freq             NaN                      2                      2   
mean        1.556236                    NaN                    NaN   
std         0.496838                    NaN                    NaN   
min         1.000000                    NaN                    NaN   
25%         1.000000                    NaN                    NaN   
50%         2.000000                    NaN                    NaN   
75%         2.000000                    NaN                    NaN   
max         2.000000                    NaN                    NaN   

        passenger_count  trip_distance    RatecodeID store_and_fwd_flag  \
count      22699.000000   22699.000000  22699.000000              22699   
unique              NaN            NaN           NaN                  2   
top                 NaN            NaN           NaN                  N   
freq                NaN            NaN           NaN              22600   
mean           1.642319       2.913313      1.043394                NaN   
std            1.285231       3.653171      0.708391                NaN   
min            0.000000       0.000000      1.000000                NaN   
25%            1.000000       0.990000      1.000000                NaN   
50%            1.000000       1.610000      1.000000                NaN   
75%            2.000000       3.060000      1.000000                NaN   
max            6.000000      33.960000     99.000000                NaN   

        PULocationID  DOLocationID  payment_type   fare_amount         extra  \
count   22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean      162.412353    161.527997      1.336887     13.026629      0.333275   
std        66.633373     70.139691      0.496211     13.243791      0.463097   
min         1.000000      1.000000      1.000000   -120.000000     -1.000000   
25%       114.000000    112.000000      1.000000      6.500000      0.000000   
50%       162.000000    162.000000      1.000000      9.500000      0.000000   
75%       233.000000    233.000000      2.000000     14.500000      0.500000   
max       265.000000    265.000000      4.000000    999.990000      4.500000   

             mta_tax    tip_amount  tolls_amount  improvement_surcharge  \
count   22699.000000  22699.000000  22699.000000           22699.000000   
unique           NaN           NaN           NaN                    NaN   
top              NaN           NaN           NaN                    NaN   
freq             NaN           NaN           NaN                    NaN   
mean        0.497445      1.835781      0.312542               0.299551   
std         0.039465      2.800626      1.399212               0.015673   
min        -0.500000      0.000000      0.000000              -0.300000   
25%         0.500000      0.000000      0.000000               0.300000   
50%         0.500000      1.350000      0.000000               0.300000   
75%         0.500000      2.450000      0.000000               0.300000   
max         0.500000    200.000000     19.100000               0.300000   

        total_amount  
count   22699.000000  
unique           NaN  
top              NaN  
freq             NaN  
mean       16.310502  
std        16.097295  
min      -120.300000  
25%         8.750000  
50%        11.800000  
75%        17.800000  
max      1200.290000

**You are interested in the relationship between payment type and the fare amount the customer pays.** One approach is to look at the average fare amount for each payment type. 

In [18]:
taxi_data.groupby('payment_type')['fare_amount'].mean()

payment_type
1    13.429748
2    12.213546
3    12.186116
4     9.913043
Name: fare_amount, dtype: float64

* The dataset represents a sample of yellow taxi trips taken in 2017, not every trip in New York City. As a result, the difference observed between fare amounts for credit card and cash payments may be due to random sampling, rather than reflecting a true difference across all trips.

* Based on the averages, customers who pay with a credit card tend to have slightly higher fare amounts compared to those paying with cash. However, this difference might not be significant—it could simply be a result of the sample selected.

* To determine whether this difference is statistically significant and not just a random finding, a hypothesis test is needed. The test will help assess whether the observed difference is large enough to be considered meaningful or if it could simply be attributed to chance within the sample. Even a small difference could have practical significance, but conducting the hypothesis test will confirm whether this difference holds true beyond just this sample of data



### Task 3. Hypothesis testing

Before you conduct your hypothesis test, consider the following questions where applicable to complete your code response:

1. Recall the difference between the null hypothesis and the alternative hypotheses. Consider your hypotheses for this project as listed below.

$H_0$: There is no difference in the average fare amount between customers who use credit cards and customers who use cash.

$H_A$: There is a difference in the average fare amount between customers who use credit cards and customers who use cash.



Your goal in this step is to conduct a two-sample t-test. Recall the steps for conducting a hypothesis test: 


1.   State the null hypothesis and the alternative hypothesis
2.   Choose a signficance level
3.   Find the p-value
4.   Reject or fail to reject the null hypothesis 



**Note:** For the purpose of this exercise, your hypothesis test is the main component of your A/B test. 

You choose 5% as the significance level and proceed with a two-sample t-test.

In [21]:
#hypothesis test, A/B test
#significance level
credit_card = taxi_data[taxi_data['payment_type']==1]['fare_amount']
cash = taxi_data[taxi_data['payment_type']==2]['fare_amount']
stats.ttest_ind(a=credit_card,b=cash,equal_var= False)

Ttest_indResult(statistic=6.866800855655372, pvalue=6.797387473030518e-12)

* The pvalue 6.797387473030518e-12 is a very small number written in scientific notation, equivalent to 0.000000000006797387473030518.It is significantly smaller than the significance level of 5%, we reject the null hypothesis.

* We conclude that there is a statistically significant difference in the average fare amount between customers who use credit cards and customers who use cash.

<img src="images/Execute.png" width="100" height="100" align=left>

## PACE: **Execute**

Consider the questions in your PACE Strategy Document to reflect on the Execute stage.

### Task 4. Communicate insights with stakeholders

*Ask yourself the following questions:*

1. What business insight(s) can you draw from the result of your hypothesis test?

* The key business insight is that encouraging customers to pay with credit cards can generate more revenue for taxi cab drivers.

2. Consider why this A/B test project might not be realistic, and what assumptions had to be made for this educational project.

* In an A/B test, the groups need to be randomly selected to make sure the results are fair and unbiased. However, in this scenario, the dataset doesn't reflect a real-world experiment where people were randomly assigned to use either cash or credit card for payment. The claim in the project is that, for the sake of the A/B test, the riders were randomly grouped into two categories (cash vs. credit card), and they were told to pay using one method. In reality, though, passengers don’t usually get told how to pay; they choose based on their own preferences and what they have available (e.g., cash or card). Some people might not carry enough cash, so they have to pay by card. So, the assumption made here is that everyone was randomly assigned, which isn’t actually how payment methods work in real life.

==> ENTER YOUR RESPONSE HERE 

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.